File: src/main/java/me/testaccount666/serversystem/commands/executables/home/admin/TabCompleterAdminHome.java
package me.testaccount666.serversystem.commands.executables.home.admin;

import me.testaccount666.serversystem.ServerSystem;
import me.testaccount666.serversystem.commands.interfaces.ServerSystemTabCompleter;
import me.testaccount666.serversystem.managers.PermissionManager;
import me.testaccount666.serversystem.userdata.OfflineUser;
import me.testaccount666.serversystem.userdata.User;
import me.testaccount666.serversystem.userdata.home.Home;
import org.bukkit.command.Command;

import java.util.List;
import java.util.Optional;

public class TabCompleterAdminHome implements ServerSystemTabCompleter {

    @Override
    public Optional<List<String>> tabComplete(User commandSender, Command command, String label, String... arguments) {
        var commandName = command.getName().substring("admin".length());

        if (commandName.equalsIgnoreCase("sethome")) return handleSetHomeCommand(commandSender, arguments);

        if (commandName.equalsIgnoreCase("deletehome")) return handleDeleteHomeCommand(commandSender, arguments);

        return handleHomeCommand(commandSender, arguments);
    }

    private Optional<List<String>> handleHomeCommand(User commandSender, String... arguments) {
        if (!PermissionManager.hasPermission(commandSender.getCommandSender(), "Commands.AdminHome.Use", false)) return Optional.empty();

        if (arguments.length <= 1) return Optional.empty();

        return handleHomeCompletion(commandSender, arguments);
    }

    private Optional<List<String>> handleSetHomeCommand(User commandSender, String... arguments) {
        if (!PermissionManager.hasPermission(commandSender.getCommandSender(), "Commands.AdminSetHome.Use", false)) return Optional.empty();
        if (arguments.length <= 1) return Optional.empty();

        return Optional.of(List.of());
    }

    private Optional<List<String>> handleDeleteHomeCommand(User commandSender, String... arguments) {
        if (!PermissionManager.hasPermission(commandSender.getCommandSender(), "Commands.AdminDeleteHome.Use", false)) return Optional.empty();
        if (arguments.length <= 1) return Optional.empty();

        return handleHomeCompletion(commandSender, arguments);
    }

    private Optional<List<String>> handleHomeCompletion(OfflineUser commandSender, String... arguments) {
        if (arguments.length != 2) return Optional.of(List.of());

        var targetUserOptional = ServerSystem.Instance.getUserManager().getUser(arguments[0]);
        if (targetUserOptional.isEmpty()) return Optional.of(List.of());

        var targetCachedUser = targetUserOptional.get();
        var targetUser = targetCachedUser.getOfflineUser();

        var homeManager = targetUser.getHomeManager();

        var potentialCompletions = homeManager.getHomes().stream().map(Home::getDisplayName).toList();
        var foundCompletions = potentialCompletions.stream().filter(completion -> completion.toLowerCase().startsWith(arguments[1].toLowerCase())).toList();

        return foundCompletions.isEmpty()? Optional.of(potentialCompletions) : Optional.of(foundCompletions);
    }
}

File: src/main/kotlin/me/testaccount666/serversystem/commands/executables/home/admin/TabCompleterAdminHome.kt
package me.testaccount666.serversystem.commands.executables.home.admin

import me.testaccount666.serversystem.ServerSystem.Companion.instance
import me.testaccount666.serversystem.commands.interfaces.ServerSystemTabCompleter
import me.testaccount666.serversystem.managers.PermissionManager.hasCommandPermission
import me.testaccount666.serversystem.userdata.User
import me.testaccount666.serversystem.userdata.UserManager
import me.testaccount666.serversystem.userdata.home.Home
import org.bukkit.command.Command

class TabCompleterAdminHome : ServerSystemTabCompleter {
    override fun tabComplete(commandSender: User, command: Command, label: String, vararg arguments: String): List<String>? {
        val commandName = command.name.substring("admin".length)

        if (commandName.equals("sethome", true)) return handleSetHomeCommand(commandSender, *arguments)

        if (commandName.equals("deletehome", true)) return handleDeleteHomeCommand(commandSender, *arguments)

        return handleHomeCommand(commandSender, *arguments)
    }

    private fun handleHomeCommand(commandSender: User, vararg arguments: String): List<String>? {
        if (!hasCommandPermission(commandSender, "AdminHome.Use", false)) return null
        if (arguments.size <= 1) return null

        return handleHomeCompletion(*arguments)
    }

    private fun handleSetHomeCommand(commandSender: User, vararg arguments: String): List<String>? {
        if (!hasCommandPermission(commandSender, "AdminHome.Set", false)) return null
        if (arguments.size <= 1) return null

        return mutableListOf()
    }

    private fun handleDeleteHomeCommand(commandSender: User, vararg arguments: String): List<String>? {
        if (!hasCommandPermission(commandSender, "AdminHome.Delete", false)) return null
        if (arguments.size <= 1) return null

        return handleHomeCompletion(*arguments)
    }

    private fun handleHomeCompletion(vararg arguments: String): List<String> {
        if (arguments.size != 2) return listOf()

        val targetCachedUser = instance.registry.getService<UserManager>().getUserOrNull(arguments[0]) ?: return listOf()
        val targetUser = targetCachedUser.offlineUser

        val homeManager = targetUser.homeManager

        val potentialCompletions = homeManager.homes.map(Home::displayName)
        return potentialCompletions.filter { it.startsWith(arguments[1], true) }
    }
}
